wasm/src/language_service.rs

```diff
@@ -1,44 +1,231 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
+use std::{future::Future, path::PathBuf, pin::Pin, sync::Arc};
+
 use crate::{diagnostic::VSDiagnostic, serializable_type};
 use js_sys::JsString;
 use qsc::{self};
+use qsc_project::{EntryType, Manifest, ManifestDescriptor};
+use qsls::{protocol::DiagnosticUpdate, JSFileEntry};
 use rustc_hash::FxHashMap;
 use serde::{Deserialize, Serialize};
 use wasm_bindgen::prelude::*;
+use wasm_bindgen_futures::JsFuture;
 
 #[wasm_bindgen]
 pub struct LanguageService(qsls::LanguageService<'static>);
 
+async fn map_js_promise<F, T>(res: JsFuture, func: F) -> T
+where
+    F: Fn(JsValue) -> T,
+{
+    let res = res.await.expect("js future shouldn't throw an exception");
+    log::trace!("asynchronous callback from wasm returned {res:?}");
+    func(res)
+}
 #[wasm_bindgen]
 impl LanguageService {
     #[wasm_bindgen(constructor)]
-    pub fn new(diagnostics_callback: DiagnosticsCallback) -> Self {
+    pub fn new(
+        diagnostics_callback: DiagnosticsCallback,
+        read_file: ReadFileCallback,
+        list_directory: ListDirectoryCallback,
+        get_manifest: GetManifestCallback,
+    ) -> Self {
+        let read_file = read_file
+            .dyn_ref::<js_sys::Function>()
+            .unwrap_or_else(|| {
+                panic!(
+                    "expected a valid JS function (read_file), received {:?}",
+                    read_file.js_typeof()
+                )
+            })
+            .clone();
+
+        let read_file = move |path_buf: PathBuf| {
+            let path_buf_string = &path_buf.to_string_lossy().to_string();
+            let path = JsValue::from_str(path_buf_string);
+            let res: js_sys::Promise = read_file
+                .call1(&JsValue::NULL, &path)
+                .expect("callback should succeed")
```

Comment:
Hmm, we really shouldn't panic on file i/o errors - we should have a strategy to recover gracefully. One easy way to hit i/o errors is if I'm deleting files in the workspace while a project recompilation is going on in the background. Especially once you hookup file watchers - bulk deletes will trigger project updates, project update will cause a mass of file reads, those file reads will fail, etc...